crates/core/tedge/src/cli/certificate/cli.rs

```diff
@@ -193,16 +196,187 @@ pub enum UploadCertCli {
     },
 }
 
-fn get_device_id_from_config(
+/// Returns the device ID from the config if no ID is provided by CLI
+/// If the provided device ID mismatches the one from the config, it returns an error
+fn get_device_id(
+    id: Option<String>,
     config: &TEdgeConfig,
+    config_location: &TEdgeConfigLocation,
     cloud: &Option<Cloud>,
-) -> Result<String, ReadError> {
-    let id = match cloud {
-        None => config.device.id(),
-        Some(Cloud::C8y(profile)) => config.c8y.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Azure(profile)) => config.az.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Aws(profile)) => config.aws.try_get(profile.as_deref())?.device.id(),
-    }?
-    .to_owned();
-    Ok(id)
+) -> Result<String, TEdgeError> {
+    let config_id = config.device_id(cloud.as_ref()).ok();
+    let writable_key = get_writable_key(cloud)?;
+    let dto = config_location
+        .load_dto_from_toml_and_env()
+        .context("failed to read tedge.toml file")?;
+
+    match id {
+        None => match config_id {
+            None => Err(anyhow!("No device ID is provided. Use `--device-id <name>` option to specify the device ID.").into()),
+            Some(config_id) => Ok(config_id.into()),
+        }
+        Some(input_id) if config_id.is_none() => Ok(input_id),
+        Some(input_id) if input_id == config_id.unwrap() => Ok(input_id),
+        Some(input_id) => match cloud.as_ref().map(<_>::into) {
+            None => Err(MismatchedDeviceId {
+                input_id,
+                config_id: config_id.unwrap().into(),
+                writable_key,
+            }),
+            Some(tedge_config::Cloud::C8y(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let c8y_dto = dto.c8y.try_get(key.as_deref(), "c8y")?;
+                if c8y_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
```

Comment:
Is this error case really useful?

- It's used only in this function
- and the error case for no device id at all is raised using  `anyhow!()`.

Reply:
Yeah, I wanted to find a better solution here. Using `anyhow!()` is definitely fine, even I used it at very first draft. But it will cause repeating the same long error message multiple times in the code. Anyway, I don't like that I had to add `MismatchedDeviceId` under `TEdgeError` as well, so using `anyhow!()` may be better.

Reply:
~~I'm now surprised that there no more conflict detection at all, when the device id is provided on the command line and in the config, but with different values. Wouldn't be better to raise an error stressing the conflict?~~

Okay after having read  https://github.com/thin-edge/thin-edge.io/pull/3326#discussion_r1923282326.